File: packages/node/src/core/signature.ts
/**
 * TC3-HMAC-SHA256 Signature Generation
 * Implements Tencent Cloud API v3 authentication
 */

import { createHash, createHmac } from 'crypto';
import type { TC3Signature } from '../types.js';

/**
 * Constants for Tencent Cloud TRTC TTS API
 */
const TTS_SERVICE = 'trtc';
const TTS_HOST = 'trtc.ai.tencentcloudapi.com';
const TTS_VERSION = '2019-07-22';

/**
 * Get current Unix timestamp in seconds
 */
function getTimestamp(): number {
  return Math.floor(Date.now() / 1000);
}

/**
 * Get date string in YYYY-MM-DD format
 */
function getDate(timestamp: number): string {
  const date = new Date(timestamp * 1000);
  const year = date.getUTCFullYear();
  const month = String(date.getUTCMonth() + 1).padStart(2, '0');
  const day = String(date.getUTCDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}

/**
 * SHA256 hash function
 */
function sha256(data: string): string {
  return createHash('sha256').update(data, 'utf8').digest('hex');
}

/**
 * HMAC-SHA256 function
 * @param data - Data to hash
 * @param key - Secret key (string or Buffer)
 * @param encoding - Output encoding ('hex' or null for Buffer)
 * @returns Hex string or raw Buffer
 */
function hmacSha256(
  data: string,
  key: string | Buffer,
  encoding: 'hex' | null
): string | Buffer {
  const hmac = createHmac('sha256', key).update(data, 'utf8');
  return encoding === 'hex' ? hmac.digest('hex') : hmac.digest();
}

/**
 * Build TC3-HMAC-SHA256 signature for Tencent Cloud API
 *
 * @param action - API action name (e.g., 'TextToSpeech')
 * @param region - Tencent Cloud region (e.g., 'ap-beijing')
 * @param payload - Request payload as JSON string
 * @param secretId - Tencent Cloud Secret ID
 * @param secretKey - Tencent Cloud Secret Key
 * @returns Signature object with authorization header and all headers
 */
export function buildTC3Signature(
  action: string,
  region: string,
  payload: string,
  secretId: string,
  secretKey: string
): TC3Signature {
  // Get timestamp and date
  const timestamp = getTimestamp();
  const date = getDate(timestamp);

  // Step 1: Build canonical request
  // Format: HTTPMethod\nURI\nQueryString\nCanonicalHeaders\nSignedHeaders\nHashedPayload
  const hashedPayload = sha256(payload);
  const canonicalHeaders = [
    `content-type:application/json; charset=utf-8`,
    `host:${TTS_HOST}`,
    `x-tc-action:${action.toLowerCase()}`
  ].join('\n') + '\n'; // Must end with \n

  const signedHeaders = 'content-type;host;x-tc-action';

  const canonicalRequest = [
    'POST',
    '/',
    '', // Empty query string
    canonicalHeaders,
    signedHeaders,
    hashedPayload
  ].join('\n');

  // Step 2: Build string to sign
  // Format: Algorithm\nTimestamp\nCredentialScope\nHashedCanonicalRequest
  const credentialScope = `${date}/${TTS_SERVICE}/tc3_request`;
  const hashedCanonicalRequest = sha256(canonicalRequest);

  const stringToSign = [
    'TC3-HMAC-SHA256',
    timestamp.toString(),
    credentialScope,
    hashedCanonicalRequest
  ].join('\n');

  // Step 3: Calculate signature
  // CRITICAL: Intermediate keys MUST be raw Buffers, not hex strings
  // kDate = HMAC-SHA256("TC3" + secretKey, date)
  // kService = HMAC-SHA256(kDate, service)
  // kSigning = HMAC-SHA256(kService, "tc3_request")
  // signature = HMAC-SHA256(kSigning, stringToSign) -> hex
  const kDate = hmacSha256(date, `TC3${secretKey}`, null) as Buffer;
  const kService = hmacSha256(TTS_SERVICE, kDate, null) as Buffer;
  const kSigning = hmacSha256('tc3_request', kService, null) as Buffer;
  const signature = hmacSha256(stringToSign, kSigning, 'hex') as string;

  // Step 4: Build Authorization header
  const authorization = `TC3-HMAC-SHA256 Credential=${secretId}/${credentialScope}, SignedHeaders=${signedHeaders}, Signature=${signature}`;

  // Return all headers required for API request
  return {
    authorization,
    headers: {
      'Content-Type': 'application/json; charset=utf-8',
      'Host': TTS_HOST,
      'X-TC-Action': action,
      'X-TC-Timestamp': timestamp.toString(),
      'X-TC-Version': TTS_VERSION,
      'X-TC-Region': region,
      'Authorization': authorization
    }
  };
}

/**
 * Build TC3 signature specifically for SSE streaming requests
 * (Same as regular signature, but may need special handling for streaming)
 */
export function buildTC3SignatureSSE(
  action: string,
  region: string,
  payload: string,
  secretId: string,
  secretKey: string
): TC3Signature {
  // SSE uses the same signature algorithm
  return buildTC3Signature(action, region, payload, secretId, secretKey);
}

File: packages/go/signature.go
package flowtts

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"
)

const (
	// TTS API constants
	ttsService = "trtc"
	ttsVersion = "2019-07-22"
	ttsHost    = "trtc.ai.tencentcloudapi.com"

	// Actions
	ttsAction       = "TextToSpeech"
	ttsActionStream = "TextToSpeechSSE"

	// Algorithm
	algorithm = "TC3-HMAC-SHA256"
)

// generateHeaders generates HTTP headers with TC3-HMAC-SHA256 signature.
func generateHeaders(secretID, secretKey, payload string, stream bool) map[string]string {
	timestamp := time.Now().Unix()
	date := time.Unix(timestamp, 0).UTC().Format("2006-01-02")

	action := ttsAction
	if stream {
		action = ttsActionStream
	}

	// Step 1: Create canonical request
	httpMethod := "POST"
	canonicalURI := "/"
	canonicalQueryString := ""
	canonicalHeaders := fmt.Sprintf("content-type:application/json\nhost:%s\n", ttsHost)
	signedHeaders := "content-type;host"

	hashedPayload := sha256Sum(payload)
	canonicalRequest := fmt.Sprintf("%s\n%s\n%s\n%s\n%s\n%s",
		httpMethod,
		canonicalURI,
		canonicalQueryString,
		canonicalHeaders,
		signedHeaders,
		hashedPayload,
	)

	// Step 2: Create string to sign
	credentialScope := fmt.Sprintf("%s/%s/tc3_request", date, ttsService)
	hashedCanonicalRequest := sha256Sum(canonicalRequest)
	stringToSign := fmt.Sprintf("%s\n%d\n%s\n%s",
		algorithm,
		timestamp,
		credentialScope,
		hashedCanonicalRequest,
	)

	// Step 3: Calculate signature
	secretDate := hmacSHA256([]byte("TC3"+secretKey), date)
	secretService := hmacSHA256(secretDate, ttsService)
	secretSigning := hmacSHA256(secretService, "tc3_request")
	signature := hex.EncodeToString(hmacSHA256(secretSigning, stringToSign))

	// Step 4: Create authorization header
	authorization := fmt.Sprintf("%s Credential=%s/%s, SignedHeaders=%s, Signature=%s",
		algorithm,
		secretID,
		credentialScope,
		signedHeaders,
		signature,
	)

	// Return headers
	headers := map[string]string{
		"Content-Type":     "application/json",
		"Host":             ttsHost,
		"X-TC-Action":      action,
		"X-TC-Version":     ttsVersion,
		"X-TC-Timestamp":   fmt.Sprintf("%d", timestamp),
		"X-TC-Region":      "", // Region not required for TTS
		"Authorization":    authorization,
	}

	return headers
}

// sha256Sum returns the SHA256 hash of the input string.
func sha256Sum(s string) string {
	h := sha256.New()
	h.Write([]byte(s))
	return hex.EncodeToString(h.Sum(nil))
}

// hmacSHA256 computes HMAC-SHA256.
func hmacSHA256(key []byte, data string) []byte {
	h := hmac.New(sha256.New, key)
	h.Write([]byte(data))
	return h.Sum(nil)
}
